### ISSUE TITLE
Supporting APL

### ISSUE THREAD

**Issue by Anteru** [pygments/pygments-migration-test#676]
(Original issue 969 created by None on 2014-03-22T21:42:31.209000+00:00)

APL is quite active now, with Dyalog APL, several online versions of the free javascript NGN APL interpreter, and the very actively maintained GNU APL. Could Pygments support APL in the future? Regards.

**Comment by Anteru** [comment]
(Original issue 969 last updated on 2014-03-24T07:23:44.994755+00:00)

(Issue automaticaly closed due to status in Bitbucket: resolved)

**Comment by Anteru** [comment]
(Original comment by baruchel on 2014-03-23T13:12:18.024808+00:00)

Here is a working lexer:

```
#!python

class APLLexer(RegexLexer):
    """
    A simple APL lexer.
    Contributed by Thomas Baruchel <baruchel@gmx.com>.
    """
    name = 'APL'
    aliases = ['apl']
    filenames = ['*.apl']

    tokens = {
        'root': [
            # Whitespace
            # ==========
            (r'\s+', Text),
            #
            # Comment
            # =======
            # '⍝' is traditional; '#' is supported by GNU APL and NGN (but not Dyalog)
            (u'[⍝#].*$', Comment.Single),
            #
            # Strings
            # =======
            (r'\'((\'\')|[^\'])*\'', String.Single),
            (r'"(("")|[^"])*"', String.Double), # supported by NGN APL
            #
            # Punctuation
            # ===========
            # This token type is used for diamond and parenthesis
            # but not for bracket and ; (see below)
            (u'[⋄◇()]', Punctuation),
            #
            # Array indexing
            # ==============
            # Since this token type is very important in APL, it is not included in
            # the punctuation token type but rather in the following one
            (r'[\[\];]', String.Regex),
            #
            # Distinguished names
            # ===================
            # following IBM APL2 standard
            (u'⎕[A-Za-zΔ∆⍙][A-Za-zΔ∆⍙_¯0-9]*', Name.Function),
            #
            # Labels
            # ======
            # following IBM APL2 standard
            # (u'[A-Za-zΔ∆⍙][A-Za-zΔ∆⍙_¯0-9]*:', Name.Label),
            #
            # Variables
            # =========
            # following IBM APL2 standard
            (u'[A-Za-zΔ∆⍙][A-Za-zΔ∆⍙_¯0-9]*', Name.Variable),
            #
            # Numbers
            # =======
            (u'¯?(0[Xx][0-9A-Fa-f]+|[0-9]*\.?[0-9]+([Ee][+¯]?[0-9]+)?|¯|∞)([Jj]¯?(0[Xx][0-9A-Fa-f]+|[0-9]*\.?[0-9]+([Ee][+¯]?[0-9]+)?|¯|∞))?', Number),
            #
            # Operators
            # ==========
            (u'[\.\\\/⌿⍀¨⍣⍨⍠⍤∘]', Name.Attribute), # closest token type
            (u'[+\-×÷⌈⌊∣|⍳?*⍟○!⌹<≤=>≥≠≡≢∊⍷∪∩~∨∧⍱⍲⍴,⍪⌽⊖⍉↑↓⊂⊃⌷⍋⍒⊤⊥⍕⍎⊣⊢⍁⍂≈⌸⍯↗]', Operator),
            #
            # Constant
            # ========
            (u'⍬', Name.Constant),
            #
            # Quad symbol
            # ===========
            (u'[⎕⍞]', Name.Variable.Global),
            #
            # Arrows left/right
            # =================
            (u'[←→]', Keyword.Declaration),
            #
            # D-Fn
            # ====
            (u'[⍺⍵⍶⍹∇:]', Name.Builtin.Pseudo),
            (r'[{}]', Keyword.Type),
        ],
    }
```

**Comment by Anteru** [comment]
(Original comment by birkenfeld on 2014-03-23T17:16:52.724276+00:00)

Thanks for the contribution. To finish this, could you turn it into a pull request complete with an example file for tests/examplefiles?

**Comment by Anteru** [comment]
(Original comment by baruchel on 2014-03-23T21:03:05.762108+00:00)

@birkenfeld. I don't know how to do the pull request you are asking for. I can perfectly send all files to anyone by mail however.

**Comment by Anteru** [comment]
(Original comment by birkenfeld on 2014-03-23T22:01:01.006280+00:00)

OK, no problem, just attach them to this issue then.

**Comment by Anteru** [comment]
(Original comment by baruchel on 2014-03-23T22:12:22.676667+00:00)

@birkenfeld Here is an example below, and both the code and the example will be attached to the issue.
```
#!APL

∇ R←M COMBIN N;D;E;F;G;P
  ⍝ Returns a matrix of every possible
  ⍝ combination of M elements from the
  ⍝ vector ⍳N.  That is, returns a
  ⍝ matrix with M!N rows and N columns.
  ⍝
  E←(⍳P←N-R←M-1)-⎕IO
  D←R+⍳P
  R←(P,1)⍴D
  P←P⍴1
 L1:→(⎕IO>1↑D←D-1)⍴0
  P←+\P
  G←+\¯1↓0,F←⌽P
  E←F/E-G
  R←(F/D),R[E+⍳⍴E;]
  E←G
  →L1
∇

∇ R←M QUICKEXP N
  ⍝ Matrix exponentiation
  B ← ⌊ 1 + 2 ⍟ N
  V ← (B ⍴ 2) ⊤ N
  L ← ⊂ M
  R ← ⊃ +.× / V / L ⊣ { L ← (⊂ A +.× A ← ↑L) , L }¨ ⍳ B-1
∇

```

**Comment by Anteru** [comment]
(Original comment by birkenfeld on 2014-03-24T07:23:20.881468+00:00)

Thanks!

**Comment by Anteru** [comment]
(Original comment by birkenfeld on 2014-03-24T07:23:44.987081+00:00)

Closes #969: add APL lexer, by Thomas Baruchel.

→ <<cset a022ac4cbe8b>>
